File: monolith/native_training/data/training_instance/python/parse_instance_ops.py
# Copyright 2022 ByteDance and/or its affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from absl import logging
import os
import struct
from typing import Dict, List, Iterable, Callable

import tensorflow as tf
from tensorflow.python.platform import resource_loader
from tensorflow.python.framework import load_library
from tensorflow.python.ops.ragged.row_partition import RowPartition, _row_partition_factory_key

from monolith.native_training.data.utils import get_slot_feature_name
from monolith.native_training.data.training_instance.python.parser_utils import \
  add_extra_parse_step, advanced_parse
from monolith.native_training.runtime.ops import gen_monolith_ops

parse_instance_ops = gen_monolith_ops


def _parse_instance_impl(
    serialized: tf.Tensor, fidv1_features: List[int], fidv2_features: List[str],
    float_features: List[str], float_feature_dims: List[int],
    int64_features: List[str], int64_feature_dims: List[int],
    string_features: List[str], string_feature_dims: List[int],
    misc_float_features: List[str], misc_float_dims: List[int],
    misc_int64_features: List[str], misc_int64_dims: List[int],
    misc_string_features: List[str], misc_string_dims: List[int],
    cc_op: Callable):
  fidv1_features = fidv1_features or []
  fidv2_features = fidv2_features or []
  float_features = float_features or []
  float_feature_dims = float_feature_dims or []
  int64_features = int64_features or []
  int64_feature_dims = int64_feature_dims or []
  string_features = string_features or []
  string_feature_dims = string_feature_dims or []
  misc_float_features = misc_float_features or []
  misc_float_dims = misc_float_dims or []
  misc_int64_features = misc_int64_features or []
  misc_int64_dims = misc_int64_dims or []
  misc_string_features = misc_string_features or []
  misc_string_dims = misc_string_dims or []
  (ragged_feature_splits, ragged_feature_values, float_feature_values,
   int64_feature_values, string_feature_values, misc_float_feature_values,
   misc_int64_feature_values, misc_string_feature_values) = cc_op(
       serialized,
       N=(len(fidv1_features) + len(fidv2_features)),
       M=len(float_features),
       O=len(int64_features),
       P=len(string_features),
       Q=len(misc_float_features),
       R=len(misc_int64_features),
       S=len(misc_string_features),
       fidv1_features=fidv1_features,
       fidv2_features=fidv2_features,
       float_features=float_features,
       float_feature_dims=float_feature_dims,
       string_features=string_features,
       string_feature_dims=string_feature_dims,
       int64_features=int64_features,
       int64_feature_dims=int64_feature_dims,
       misc_float_features=misc_float_features,
       misc_float_dims=misc_float_dims,
       misc_int64_features=misc_int64_features,
       misc_int64_dims=misc_int64_dims,
       misc_string_features=misc_string_features,
       misc_string_dims=misc_string_dims,
   )
  ragged_keys = [get_slot_feature_name(slot_id) for slot_id in fidv1_features
                ] + fidv2_features

  ragged_values = []
  for values, row_splits in zip(ragged_feature_values, ragged_feature_splits):
    row_partition = RowPartition(
        row_splits,
        # value_rowids=
        # nrows=
        #
        # TODO(zhuoran): Besides the "value" and "split" parsed from proto above,
        # precompute other two encodings "value_rowids" & "nrows" in Fountain also,
        # so that we could construct the ragged tensor with 4 precomputed encodings,
        # and would not need to recompute them later at training period again.
        internal=_row_partition_factory_key
        # Currently, we just compute and cache value_rowids and nrows here:
    ).with_precomputed_value_rowids().with_precomputed_nrows()
    ragged_values.append(tf.RaggedTensor(values, row_partition, internal=True))

  float_keys = float_features
  int64_keys = int64_features
  string_keys = string_features
  return dict(
      zip(
          ragged_keys + float_keys + int64_keys + string_keys +
          misc_float_features + misc_int64_features + misc_string_features,
          ragged_values + float_feature_values + int64_feature_values +
          string_feature_values + misc_float_feature_values +
          misc_int64_feature_values + misc_string_feature_values))


def parse_instances2(serialized: tf.Tensor,
                     fidv1_features: List[int] = None,
                     fidv2_features: List[str] = None,
                     float_features: List[str] = None,
                     float_feature_dims: List[int] = None,
                     int64_features: List[str] = None,
                     int64_feature_dims: List[int] = None,
                     string_features: List[str] = None,
                     string_feature_dims: List[int] = None,
                     misc_float_features: List[str] = None,
                     misc_float_dims: List[int] = None,
                     misc_int64_features: List[str] = None,
                     misc_int64_dims: List[int] = None,
                     misc_string_features: List[str] = None,
                     misc_string_dims: List[int] = None):
  """从序列化的instance Tensor中解析instance
  
  Args:
    varient_tensor (:obj:`Tensor`): 输入数据
    fidv1_features (:obj:`List[int]`): 在Instance中, fidv1_features是平铺的, 所以用slot指定, 可以是部分slot
    fidv2_features (:obj:`List[str]`): 在Instance中, fidv2_features存放于feature中, 可以用名字指定, 可以是部分特征名
    float_features (:obj:`List[str]`): 在Instance中, 连续特征存于feature中, 可以用名字指定, 可以是部分特征名
    float_feature_dims (:obj:`List[int]`): 连续特征的维度, `float_feature_dims`的长度要与`float_features`一致
    int64_features (:obj:`List[str]`): 在Instance中, int64特征(非FID)存于feature中, 可以用名字指定, 可以是部分特征名
    int64_feature_dims (:obj:`List[int]`): int64特征的维度, `int64_feature_dims`的长度要与`int64_features`一致
    string_features (:obj:`List[str]`): 在Instance中, syting特征存于feature中, 可以用名字指定, 可以是部分特征名
    string_feature_dims (:obj:`List[int]`): string特征的维度, `string_feature_dims`的长度要与`string_features`一致
    misc_float_features (:obj:`List[str]`): 在LineId中, float字段, 用名字指定, 可以有多个
    misc_float_dims (:obj:`List[int]`): 在LineId中, float字段维度, `misc_float_dims`的长度要与`misc_float_features`一致
    misc_int64_features (:obj:`List[str]`): 在LineId中, int64字段, 用名字指定, 可以有多个
    misc_int64_dims (:obj:`List[int]`): 在LineId中, int64字段维度, `misc_int64_dims`的长度要与`misc_int64_features`一致
    misc_string_features (:obj:`List[str]`): 在LineId中, string字段, 用名字指定, 可以有多个
    misc_string_dims (:obj:`List[str]`): 在LineId中, string字段维度, `misc_string_dims`的长度要与`misc_string_features`一致

  Returns:
    Dict[str, Tensor] 解析出特征名到特征的字典
  
  """

  return _parse_instance_impl(
      serialized, fidv1_features, fidv2_features, float_features,
      float_feature_dims, int64_features, int64_feature_dims, string_features,
      string_feature_dims, misc_float_features, misc_float_dims,
      misc_int64_features, misc_int64_dims, misc_string_features,
      misc_string_dims, parse_instance_ops.monolith_parse_instances)


def parse_instances(serialized: tf.Tensor,
                    fidv1_features: List[int] = None,
                    fidv2_features: List[str] = None,
                    float_features: List[str] = None,
                    float_feature_dims: List[int] = None,
                    int64_features: List[str] = None,
                    int64_feature_dims: List[int] = None,
                    string_features: List[str] = None,
                    string_feature_dims: List[int] = None,
                    misc_float_features: List[str] = ['sample_rate'],
                    misc_int64_features: List[str] = ['req_time', 'uid'],
                    misc_string_features: List[str] = None,
                    misc_repeated_float_features: List[str] = ['label'],
                    misc_repeated_float_dims: List[int] = None,
                    misc_repeated_int64_features: List[str] = None,
                    misc_repeated_int64_dims: List[int] = None,
                    misc_repeated_string_features: List[str] = None,
                    misc_repeated_string_dims: List[str] = None):
  """从序列化的instance Tensor中解析instance, 但参数较多, 请使用`parse_instances2`
  
  Args:
    varient_tensor (:obj:`Tensor`): 输入数据
    fidv1_features (:obj:`List[int]`): 在Instance中, fidv1_features是平铺的, 所以用slot指定, 可以是部分slot
    fidv2_features (:obj:`List[str]`): 在Instance中, fidv2_features存放于feature中, 可以用名字指定, 可以是部分特征名
    float_features (:obj:`List[str]`): 在Instance中, 连续特征存于feature中, 可以用名字指定, 可以是部分特征名
    float_feature_dims (:obj:`List[int]`): 连续特征的维度, `float_feature_dims`的长度要与`float_features`一致
    int64_features (:obj:`List[str]`): 在Instance中, int64特征(非FID)存于feature中, 可以用名字指定, 可以是部分特征名
    int64_feature_dims (:obj:`List[int]`): int64特征的维度, `int64_feature_dims`的长度要与`int64_features`一致
    string_features (:obj:`List[str]`): 在Instance中, syting特征存于feature中, 可以用名字指定, 可以是部分特征名
    string_feature_dims (:obj:`List[int]`): string特征的维度, `string_feature_dims`的长度要与`string_features`一致
    misc_float_features (:obj:`List[str]`): 在LineId中, 非repeated float字段, 用名字指定, 可以有多个
    misc_int64_features (:obj:`List[str]`): 在LineId中, 非repeated int64字段, 用名字指定, 可以有多个
    misc_string_features (:obj:`List[str]`): 在LineId中, 非repeated string字段, 用名字指定, 可以有多个
    misc_repeated_float_features (:obj:`List[str]`): 在LineId中, repeated float字段, 用名字指定, 可以有多个
    misc_repeated_float_dims (:obj:`List[int]`): 在LineId中, repeated float字段维度, `misc_repeated_float_dims`的长度要与`misc_repeated_float_features`一致
    misc_repeated_int64_features (:obj:`List[str]`): 在LineId中, repeated int64字段, 用名字指定, 可以有多个
    misc_repeated_int64_dims (:obj:`List[int]`): 在LineId中, repeated int64字段维度, `misc_repeated_int64_dims`的长度要与`misc_repeated_int64_features`一致
    misc_repeated_string_features (:obj:`List[str]`): 在LineId中, repeated string字段, 用名字指定, 可以有多个
    misc_repeated_string_dims (:obj:`List[str]`): 在LineId中, repeated string字段维度, `misc_repeated_string_dims`的长度要与`misc_repeated_string_features`一致

  Returns:
    Dict[str, Tensor] 解析出特征名到特征的字典
  
  """

  fidv1_features = fidv1_features or []
  fidv2_features = fidv2_features or []
  float_features = float_features or []
  float_feature_dims = float_feature_dims or []
  int64_features = int64_features or []
  int64_feature_dims = int64_feature_dims or []
  string_features = string_features or []
  string_feature_dims = string_feature_dims or []
  misc_float_features = misc_float_features or []
  misc_float_feature_dims = [1] * len(misc_float_features)
  misc_int64_features = misc_int64_features or []
  misc_int64_feature_dims = [1] * len(misc_int64_features)
  misc_string_features = misc_string_features or []
  misc_string_features_dims = [1] * len(misc_string_features)
  misc_repeated_float_features = misc_repeated_float_features or []
  misc_repeated_float_dims = misc_repeated_float_dims or [1] * len(
      misc_repeated_float_features)
  misc_repeated_int64_features = misc_repeated_int64_features or []
  misc_repeated_int64_dims = misc_repeated_int64_dims or [1] * len(
      misc_repeated_int64_features)
  misc_repeated_string_features = misc_repeated_string_features or []
  misc_repeated_string_dims = misc_repeated_string_dims or [1] * len(
      misc_repeated_string_features)

  features = parse_instances2(
      serialized, fidv1_features, fidv2_features, float_features,
      float_feature_dims, int64_features, int64_feature_dims, string_features,
      string_feature_dims, misc_float_features + misc_repeated_float_features,
      misc_float_feature_dims + misc_repeated_float_dims,
      misc_int64_features + misc_repeated_int64_features,
      misc_int64_feature_dims + misc_repeated_int64_dims,
      misc_string_features + misc_repeated_string_features,
      misc_string_features_dims + misc_repeated_string_dims)
  for key in misc_float_features + misc_int64_features:
    features[key] = tf.reshape(features[key], [-1])

  return features


# This is mainly for test purpose, DO NOT use it directly.
monolith_raw_parse_instance = parse_instance_ops.MonolithRawParseInstance

File: monolith/native_training/data/training_instance/cc/parse_instance_ops.cc
// Copyright 2022 ByteDance and/or its affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"

namespace tensorflow {
namespace monolith_tf {
namespace {

bool IsUnsetHandle(shape_inference::DimensionHandle handle) {
  return handle.Handle() == 0;
}

shape_inference::ShapeHandle GetBatched1D(
    shape_inference::InferenceContext *ctx,
    shape_inference::DimensionHandle batch_size, int dim) {
  if (IsUnsetHandle(batch_size)) {
    return ctx->Vector(dim);
  } else {
    return ctx->Matrix(batch_size, dim);
  }
}

Status SetParseInstanceShape(shape_inference::InferenceContext *ctx,
                             shape_inference::DimensionHandle batch_size) {
  int offset = 0;

  // Ragged tensor
  int n;
  TF_RETURN_IF_ERROR(ctx->GetAttr("N", &n));
  for (int i = 0; i < n; ++i) {
    if (IsUnsetHandle(batch_size)) {
      ctx->set_output(offset + i, ctx->Vector(2));
      continue;
    }
    int batch_size_value = ctx->Value(batch_size);
    if (batch_size_value == shape_inference::InferenceContext::kUnknownDim) {
      ctx->set_output(offset + i, ctx->Vector(ctx->UnknownDim()));
    } else {
      ctx->set_output(offset + i, ctx->Vector(batch_size_value + 1));
    }
  }
  offset += n;

  for (int i = 0; i < n; ++i) {
    ctx->set_output(offset + i, ctx->Vector(ctx->UnknownDim()));
  }
  offset += n;

  // float tensor
  int m;
  TF_RETURN_IF_ERROR(ctx->GetAttr("M", &m));
  std::vector<int> float_feature_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("float_feature_dims", &float_feature_dims));
  for (int i = 0; i < m; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, float_feature_dims[i]));
  }
  offset += m;

  // int64 tensor
  int o;
  TF_RETURN_IF_ERROR(ctx->GetAttr("O", &o));
  std::vector<int> int64_feature_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("int64_feature_dims", &int64_feature_dims));
  for (int i = 0; i < o; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, int64_feature_dims[i]));
  }
  offset += o;

  // string tensor
  int p;
  TF_RETURN_IF_ERROR(ctx->GetAttr("P", &p));
  std::vector<int> string_feature_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("string_feature_dims", &string_feature_dims));
  for (int i = 0; i < p; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, string_feature_dims[i]));
  }
  offset += p;

  // misc_feature_float
  int q;
  TF_RETURN_IF_ERROR(ctx->GetAttr("Q", &q));
  std::vector<int> misc_float_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("misc_float_dims", &misc_float_dims));
  for (int i = 0; i < q; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, misc_float_dims[i]));
  }
  offset += q;

  // misc_feature_int64
  int r;
  TF_RETURN_IF_ERROR(ctx->GetAttr("R", &r));
  std::vector<int> misc_int64_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("misc_int64_dims", &misc_int64_dims));
  for (int i = 0; i < r; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, misc_int64_dims[i]));
  }
  offset += r;

  // misc_feature_string
  int s;
  TF_RETURN_IF_ERROR(ctx->GetAttr("S", &s));
  std::vector<int> misc_string_dims;
  TF_RETURN_IF_ERROR(ctx->GetAttr("misc_string_dims", &misc_string_dims));
  for (int i = 0; i < s; ++i) {
    ctx->set_output(offset + i,
                    GetBatched1D(ctx, batch_size, misc_string_dims[i]));
  }
  offset += s;

  return Status::OK();
}

// We use fid_features for FIDV1 key and str_features for FIDv2 keys.
// In fid v1, we use the slot whitelist, and in fid v2, we use the
// feature_name
// whitelist.
REGISTER_OP("MonolithParseInstances")
    .Input("serialized: string")
    .Output("ragged_feature_splits: N * int64")
    .Output("ragged_feature_values: N * int64")
    .Output("float_feature_values: M * float32")
    .Output("int64_feature_values: O * int64")
    .Output("string_feature_values: P * string")
    .Output("misc_float_feature_values: Q * float32")
    .Output("misc_int64_feature_values: R * int64")
    .Output("misc_string_feature_values: S * string")
    .Attr("N: int >= 0")
    .Attr("M: int >= 0")
    .Attr("O: int >= 0")
    .Attr("P: int >= 0")
    .Attr("Q: int >= 0")
    .Attr("R: int >= 0")
    .Attr("S: int >= 0")
    .Attr("fidv1_features: list(int)")
    .Attr("fidv2_features: list(string)")
    .Attr("float_features: list(string)")
    .Attr("float_feature_dims: list(int)")
    .Attr("int64_features: list(string)")
    .Attr("int64_feature_dims: list(int)")
    .Attr("string_features: list(string)")
    .Attr("string_feature_dims: list(int)")
    .Attr("misc_float_features: list(string)")
    .Attr("misc_float_dims: list(int)")
    .Attr("misc_int64_features: list(string)")
    .Attr("misc_int64_dims: list(int)")
    .Attr("misc_string_features: list(string)")
    .Attr("misc_string_dims: list(int)")
    .SetDoNotOptimize()  // Source dataset ops must disable constant folding.
    .SetShapeFn([](shape_inference::InferenceContext *ctx) {
      shape_inference::DimensionHandle batch_size = ctx->Dim(ctx->input(0), 0);
      return SetParseInstanceShape(ctx, batch_size);
    });

REGISTER_OP("MonolithRawParseInstance")
    .Attr("T: list(type)")
    .Input("serialized: string")
    .Output("tensors : T")
    .Attr("fidv1_features: list(int) = []")
    .Attr("fidv2_features: list(string) = []")
    .Attr("float_features: list(string) = []")
    .Attr("float_feature_dims: list(int) = []")
    .Attr("int64_features: list(string) = []")
    .Attr("int64_feature_dims: list(int) = []")
    .Attr("string_features: list(string) = []")
    .Attr("string_feature_dims: list(int) = []")
    .Attr("misc_float_features: list(string) = []")
    .Attr("misc_float_dims: list(int) = []")
    .Attr("misc_int64_features: list(string) = []")
    .Attr("misc_int64_dims: list(int) = []")
    .Attr("misc_string_features: list(string) = []")
    .Attr("misc_string_dims: list(int) = []")
    .Attr("collapse_batch_dim: bool = false")
    .Attr("fid_output_type: {'REGULAR', 'CONCAT'} = 'REGULAR'")
    .SetDoNotOptimize()  // Source dataset ops must disable constant folding.
    .SetShapeFn([](shape_inference::InferenceContext *ctx) {
      return Status::OK();
    });

}  // namespace
}  // namespace monolith_tf
}  // namespace tensorflow
